server/src/main/java/org/elasticsearch/search/query/TopDocsCollectorContext.java

```diff
@@ -357,6 +364,20 @@ static int shortcutTotalHitCount(IndexReader reader, Query query) throws IOExcep
                 count += context.reader().docFreq(term);
             }
             return count;
+        } else if (query.getClass() == DocValuesFieldExistsQuery.class && reader.hasDeletions() == false) {
+            final String field = ((DocValuesFieldExistsQuery) query).getField();
+            int count = 0;
+            for (LeafReaderContext context : reader.leaves()) {
+                FieldInfos fieldInfos = context.reader().getFieldInfos();
+                FieldInfo fieldInfo = fieldInfos.fieldInfo(field);
+                if (fieldInfo.getPointIndexDimensionCount() > 0) {
+                    count += PointValues.getDocCount(context.reader(), field);
+                } else if (fieldInfo.getIndexOptions() != IndexOptions.NONE) {
+                    Terms terms = context.reader().terms(field);
+                    count += terms.getDocCount();
+                }
```

Comment:
We need an `else` statement to handle fields that have doc values but are not indexed. In that case the shortcut doesn't work.

Reply:
added